### PR TITLE
Impl signing for minting/burning in new tx flow

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] cleaning old minting/burning enpoint
- [x] add postPolicyKey endpoint used when the already existent wallet does not have policy xpub
- [x] add signing
- [x] Adjust fee calculation
- [x] show signing in integration testing
- [x] add optional param to POST policy key
- [x] unify construct/decode mint/burn responses 

### Comments
Minting/burning working without timelocks at this moment - needs validation integral fixing to enable this - coming in the next PR

I want to have construct/decode responses:
```
{
  "tokens": [
    {
      "policy_id": "0fc891fb7368d3b7c7b88815c203fda0d6862b0f1d797222672e91fe",
      "policy_script": "policy_vkh1ecqfpeunygcgl4tqqycn86x6mhna99vsrm5pk22vq8dvxfq3htn", 
      "assets" : [
           { "asset_name": "1111",
             "quantity": 1,
             "fingerprint": "asset17jwl24z89xcgljv57fe4de558zvwl6k9h0hf22" 
           }, 
           { "asset_name": "2222",
             "quantity": 10,
             "fingerprint": "asset17jwl24z89xcgljv57fe4de558zvwl66666666" 
           }, 
      ]
    }
  ],
  "wallet_policy_key_index": "0H",
  "wallet_policy_key_hash": "policy_vkh1ecqfpeunygcgl4tqqycn86x6mhna99vsrm5pk22vq8dvxfq3htn"
}
```
### Issue Number
adp-955
adp-1542

### CODE REVIEW

File: specifications/api/swagger.yaml [hunk 1]
```diff
@@ -6096,6 +6045,24 @@ paths:
       responses: *responsesGetShelleyKey
 
   /wallets/{walletId}/policy-key:
+    post:
+      operationId: postPolicyKey
+      tags: ["Transactions"]
```
Comment (piotr-iohk) [inline specifications/api/swagger.yaml]: I think it should go to "Keys"

File: specifications/api/swagger.yaml [hunk 2]
```diff
@@ -6096,6 +6045,24 @@ paths:
...
+      summary: Create
```
Comment (piotr-iohk) [inline specifications/api/swagger.yaml]:
```suggestion
      summary: Create Policy Key
```

File: specifications/api/swagger.yaml [hunk 3]
```diff
@@ -6096,6 +6045,24 @@ paths:
...
+      description: |
+        <p align="right">status: <strong>stable</strong></p>
+
+        Create policy key for the wallet. Used for public account key derived
+        wallets or already instantiated mnemonic wallets to set policy public key
```
Comment (piotr-iohk) [inline specifications/api/swagger.yaml]:
```suggestion
        Create policy key for the wallet. 
        
        In order to be able to mint/burn assets with `POST Construct` endpoint there needs to be 
        a policy key set for the wallet. Invoking this endpoint would be required for all wallets instantiated
        before introducing mint/burn feature prior to making a mint/burn transaction from them.

```

Comment (piotr-iohk) [reply specifications/api/swagger.yaml]: I guess that this endpoint now invalidates https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getPolicyKey... I mean there's basically no need for it, right?

Also `postPolicyKey` returns actually a hash (`policy_vkh`), perhaps we should add a query parameter hash=true here and return `policy_vk` or `policy_vkh` based on it?

Comment (paweljakubas) [reply specifications/api/swagger.yaml]: now, for the consistency I added optional parameter like in getPolicyKey.
I would advocate for maintaining `getPolicyKey` in line with `postPolicyKey`. Could be helpful for consumers of the API, also POST writes to DB, and GET is reading

Comment (piotr-iohk) [reply specifications/api/swagger.yaml]: Ok, sounds good to leave `getPolicyKey`, but perhaps for `postPolicyKey` the `POST` should be actually `PUT`... it doesn't create new one every time, but sets one if doesn't exists (or overwrites the old one, but it is always the same if I'm not mistaken)

**Review by sevanspowell — COMMENTED** [review]
Thanks for this work @paweljakubas. This kind of work which involves making changes from the lowest level to the highest is difficult.

I've read through everything and it seems sensible. However, just like the last PR I reviewed it is hard to give proper review. 

For me, being able to read and understand the tests is more important than the code and it's performance characteristics. In my review I can't assert that:
  - All edge cases are being tested.
  - Tests do indeed test what they're trying to test.

I have two requests:
1. Is it possible to structure the code so that it can be property tested? I'd be really interested in property tests that cover the edge cases of common attack vectors like burning tokens you don't have, minting and burning at the same time, etc.
    - Structuring the code such that it can be property tested may make it easier to review too.
2. Could we raise the level of abstraction in the integration tests? It's becoming increasingly difficult to untangle the core of the integration tests from the code (i.e. "with this input, I expect this output").
    - Even just some high-level functions (no need to make a new DSL) could help here.

I understand you're just working with what you've got, and these aren't issues you've introduced, but I can't give meaningful review without a clearer relationship between the code and the proof that code works.

**Comment by paweljakubas** [comment]
hi @sevanspowell ,

Thanks for looking at the PR.
Several things were solved here:
1. adding postPolicyKey endpoint - needed for already existing wallets to participate in minting/burning
2. work improving what decodeTransaction emit when minting/burning is included.
3.  signing. and here we have several parts:
- detect during signTransaction that there is minting/burning with *our* policy_keys - here we learn about our policy key by callign getPolicyKey
- if we need to add the policy key private key, then we derive it and sneak it in `addVkWitnesses`
- on shelley level we use this key and create shelley witness, here we try not to be redundant if possible
- some changes were needed in also fee skeleton
4. integration tests can't be much simpler imho. I have now two backbones, one for minting, one for burning. And they are reused in several tests, for example, when showing burning the minting part is reused. And each part just checks all expectations from construct - signing - submitting - decode and finally when looking at balance and seeing presence of assets or absence.

I plan to un-pend the pending tests in the forthcoming PR. I will also think about testing this during property testing. The next PRs will be very focused. This one is big as I needed to prove that minting/burning signing really works in integration testing

**Comment by piotr-iohk** [comment]
>Is it possible to structure the code so that it can be property tested? I'd be really interested in property tests that cover the edge cases of common attack vectors like burning tokens you don't have, minting and burning at the same time, etc.
Structuring the code such that it can be property tested may make it easier to review too.

If this is possible that would be awesome. I think testing all cases on integration level is a bit of overkill so if there are any cases that could be verified with property tests that would be great. Then integration tests can be more about e2e flow rather than edge cases.

Just to note, besides the integration tests that are pending here and cover some of the edge cases (too big amount when minting, to long asset_name when minting) I have discovered following issues while working on e2e tests in #3207:
 - minting to `receiving_address` doesn't work. Minted tokens always land on source wallet balance,
 - minting and burning in single transaction doesn't work, submitting returns:
```
{
   "code":"created_invalid_transaction",
   "message":"The submitted transaction was rejected by the local node. Here's an error message that may help with debugging: TxValidationErrorInMode (ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (ScriptWitnessNotValidatingUTXOW (fromList [ScriptHash \"971551a411b40f2ab6636e94e5404e0ef496b58cf2074effc957e4b5\",ScriptHash \"be37b36b712c083bc6e4fdf81f665e315131c3057ec552ab62f20ea8\"])))])) AlonzoEraInCardanoMode"
}

```

**Comment by paweljakubas** [comment]
@sevanspowell @piotr-iohk @jonathanknowles yes, we can try to address all edge cases in next PR. This PR is adding the backbone of signing, plus e2e integration tests showing he case. This is first interaction out of three needed to have it in production. I am in favor of merging it, and then (1) improve edge cases via adding more integration tests and property tests, (2) fixing validity interval to enable timelocks. Please take notice that this affects only new tx workflow which is STILL experimental. Otherwise, we could spend many days getting (1) married with some signing/fee/server/wallet bits. Let's merge this and address (1) and (2) in next PRs and move on.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs [hunk 4]
```diff
@@ -3205,66 +3215,19 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
...
-        rDecodedTx <- request @(ApiDecodedTransaction n) ctx
-            (Link.decodeTransaction @'Shelley wa) Default decodePayload
-        verify rDecodedTx
-            [ expectResponseCode HTTP.status202
-            , expectField #assetsMinted (`shouldBe` activeAssetsInfo)
-            , expectField #assetsBurned (`shouldBe` inactiveAssetsInfo)
-            ]
+    it "TRANS_NEW_CREATE_10e - Minting assets with timelocks" $
+        \ctx -> runResourceT $ do
+        liftIO $ pendingWith "Should work when interval validity is addressed"
```
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs]: @paweljakubas 

Is it possible to add a "TODO" comment here, to link to the issue in JIRA where the internal validity issue is described?

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 5]
```diff
@@ -2474,19 +2446,49 @@ decodeTransaction ctx (ApiT wid) (ApiSerialisedTransaction (ApiT sealed)) = do
...
+        , policyScript = ApiT script
+        , assets = NE.map (toTokenAsset policy) tokens
+        }
+
+    toApiTokens (TokenMapWithScripts tokenMap scriptMap) =
+        if tokenMap == TokenMap.empty then
+            []
+        else
+           map fromIdScriptAssets $
+           toIdScriptAssets scriptMap tokenMap
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: Just curious if this would also work:

```suggestion
        map fromIdScriptAssets $
        toIdScriptAssets scriptMap tokenMap
```

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 6]
```diff
@@ -2474,19 +2446,49 @@ decodeTransaction ctx (ApiT wid) (ApiSerialisedTransaction (ApiT sealed)) = do
...
-            Just xpub ->
-                Just $ uncurry ApiPolicyKey (computeKeyPayload (Just True) xpub)
-            Nothing ->
-                Nothing
+        { tokens = toApiTokens tokenWithScripts
+        , walletPolicyKeyHash =
+            uncurry ApiPolicyKey . computeKeyPayload (Just True) <$>
+            includePolicyKeyInfo tokenWithScripts xpubM
+        , walletPolicyKeyIndex =
+            (const policyIx) <$> includePolicyKeyInfo tokenWithScripts xpubM
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]:
```suggestion
            policyIx <$ includePolicyKeyInfo tokenWithScripts xpubM
```

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 7]
```diff
@@ -3362,15 +3364,22 @@ getPolicyKey ctx (ApiT wid) hashed = do
...
-mintBurnAssets _ctx (ApiT _wid) _body = liftHandler $ throwE $
-    ErrMintBurnNotImplemented
-    "Minting and burning are not supported yet - this is just a stub"
+    -> Maybe Bool
+    -> ApiPostPolicyKeyData
+    -> Handler ApiPolicyKey
+postPolicyKey ctx (ApiT wid) hashed apiPassphrase =
+    withWorkerCtx @_ @s @ShelleyKey ctx wid liftE liftE $ \wrk -> do
+        k <- liftHandler $ W.writePolicyPublicKey @_ @s @n wrk wid pwd
+        pure $ uncurry ApiPolicyKey (computeKeyPayload hashed (getRawKey k))
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:3378]: You've already got one `$`, so perhaps go all the way? 😄 

```suggestion
        pure $ uncurry ApiPolicyKey $ computeKeyPayload hashed $ getRawKey k
```

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 8]
```diff
@@ -4384,13 +4389,18 @@ instance IsServerError ErrReadPolicyPublicKey where
...
                 ]
         ErrReadPolicyPublicKeyAbsent ->
-            apiError err403 InvalidWalletType $ T.unwords
+            apiError err403 MissingPolicyPublicKey $ T.unwords
                 [ "It seems the wallet lacks a policy public key. It's"
                 , "therefore not possible to create a minting or burning"
-                , "transaction. Please restore the wallet from a mnemonic"
-                , "phrase instead of an account public key."
+                , "transaction. Please invoke POST call on endpoint"
+                , "/wallets/{walletId}/policy-key to set it."
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]:
```suggestion
                , "transaction. Please POST to endpoint"
                , "/wallets/{walletId}/policy-key to set it."
```

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 9]
```diff
@@ -1223,17 +1219,32 @@ data ApiAnyCertificate n =
     deriving (Eq, Generic, Show)
     deriving anyclass NFData
 
-data ApiPolicyScript = ApiPolicyScript
+newtype ApiPostPolicyKeyData = ApiPostPolicyKeyData
+    { passphrase :: (ApiT (Passphrase "raw"))
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]:
```suggestion
    { passphrase :: ApiT (Passphrase "raw")
```

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 10]
```diff
@@ -1245,8 +1256,8 @@ data ApiDecodedTransaction (n :: NetworkDiscriminant) = ApiDecodedTransaction
     , outputs :: ![ApiTxOutputGeneral n]
     , collateral :: ![ApiTxInputGeneral n]
     , withdrawals :: ![ApiWithdrawalGeneral n]
-    , assetsMinted :: !ApiAssetMintBurn
-    , assetsBurned :: !ApiAssetMintBurn
+    , mint :: !ApiAssetMintBurn
+    , burn :: !ApiAssetMintBurn
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Types.hs:1260]: I like this: it's much shorter, but still clear! 👍🏻

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 11]
```diff
@@ -1955,6 +1946,9 @@ instance Arbitrary (Quantity "percent" Double) where
     shrink _ = [Quantity 0.0]
     arbitrary = Quantity <$> choose (0,100)
 
+instance Arbitrary ApiPostPolicyKeyData where
+    arbitrary = ApiPostPolicyKeyData <$> arbitrary
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: You can save a bit of boilerplate here:
```suggestion
deriving instance Arbitrary ApiPostPolicyKeyData
```

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 12]
```diff
@@ -1223,17 +1219,32 @@ data ApiAnyCertificate n =
...
     deriving anyclass NFData
 
-data ApiPolicyScript = ApiPolicyScript
+newtype ApiPostPolicyKeyData = ApiPostPolicyKeyData
+    { passphrase :: (ApiT (Passphrase "raw"))
+    }
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiTokenAsset = ApiTokenAsset
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Perhaps call this `ApiTokenAmountFingerprint`?

I feel the juxtaposition of `Token` and `Asset` might be confusing here.

(And it there's the fact that this also includes an amount and a fingerprint.)

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 13]
```diff
@@ -463,6 +472,16 @@ signTransaction networkId resolveRewardAcct resolveAddress resolveInput (body, w
 
         certs = cardanoCertKeysForWitnesses $ Cardano.txCertificates bodyContent
 
+        mintBurnScripts =
+            let (_, toMint, toBurn, _) = fromCardanoTx $
+                    Cardano.makeSignedTransaction wits body
+            in L.nub $ getScripts toMint <> getScripts toBurn
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: Question: why is it important to use `L.nub` here?

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 14]
```diff
@@ -477,6 +496,10 @@ signTransaction networkId resolveRewardAcct resolveAddress resolveInput (body, w
     mkWdrlCertWitness a = do
         mkShelleyWitness body <$> resolveRewardAcct a
 
+    mkPolicyWitness :: KeyHash -> Maybe (Cardano.KeyWitness era)
+    mkPolicyWitness a = do
+        mkShelleyWitness body <$> resolvePolicyKey a
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
    mkPolicyWitness a =
        mkShelleyWitness body <$> resolvePolicyKey a
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @paweljakubas 

I've left some comments and questions!

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 8 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Fixed in 2a92320c1a0506000d22213a01d2ce8b71387cea.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 11 above)
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Fixed in ba75be5e521293bb6adeb392c92fb52f0c625a52.

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 12 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: Fixed in 6951314468399a163fa58739c5b036bf1cbcb3b4.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs]: yes, we have validity interval ticket for that. It is the subject of third step to conclude minting/burning thorough support. I should start doing this next week

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 13 above)
Comment (paweljakubas) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: because, we can have scripts having the same policyXPub repeated and it is enough to deliver ONE wits for that

Comment (paweljakubas) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: example:
```
any [ all [cosigner#0, after slot1], all [cosigner#0, before slot2]
```
One wits is needed here. the one produced with signing key for cosigner#0. See it?

Comment (jonathanknowles) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: Okay, I've added a comment to the code:

90499017b73e990491cd7dbf0ffc735ee828955c

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs]: Okay, I've added a TODO that links to https://input-output.atlassian.net/browse/ADP-1193 at the top of this test!

Fixed in: 230d8c1b56538b6063fd45d7c4545995a317bbea.

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 6 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Fixed in 8c03c468cbf56a891d7b5c229ba9e468dcfc9260.

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 9 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: Fixed in 7076da5d9e70a8337a0ad7c822c2e49866714431.

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 14 above)
Comment (jonathanknowles) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: Fixed in 7076da5d9e70a8337a0ad7c822c2e49866714431.

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 5 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Fixed in 1c4da7cd740b501655847f312ad170f91d345915.

**Review by jonathanknowles — APPROVED** [review]
Hi Pawel

Many thanks for creating this PR!

I think it's going in the right direction. However, I do share the concerns that @sevanspowell  expressed above, namely that:

> Is it possible to structure the code so that it can be property tested? I'd be really interested in property tests that cover the edge cases of common attack vectors like burning tokens you don't have, minting and burning at the same time, etc.

I also feel we should invest more time in writing some focussed property tests, in addition to the integration tests you've already written.

I feel the main advantage would be that we can generate hundreds of random examples, rather than the very small number we're able to test in our integration test framework. We'd also gain a shorter editing-testing iteration cycle, which means faster feedback when things break. Lastly, it might help us to break code up into smaller, clearer chunks, which would answer Sam's point about making the code easier to review.

At the end of the day, as you pointed out, we do still need the integration tests to prove that the system works from end-to-end, but I feel this should be more of a "sanity check" rather than the main thrust of our test effort.

Perhaps we can add more property testing in the next PR? In particular, I'm curious to verify what happens in the following edge cases, in addition to what Sam already mentioned:

- minting and burning in the same transaction (disjoint sets of assets)
- minting and burning in the same transaction (intersecting sets of assets: this might provoke some failures)
- attempting to burn assets that are not available
- attempting to mint or burn large numbers of assets (so that the map is very large: I wonder whether we'll hit upon some kind of limit).

I suspect there are also lots of other things we can test.

Anyway, I'm happy to approve, but I do think it would also be good to also get a review from someone who is more knowledgeable about what kind of design we really want for the new transaction API. (I can only review this part very superficially!) Perhaps @rvl has an opinion here.

Thanks again!

Jonathan

**Comment by paweljakubas** [comment]
thanks @jonathanknowles a lot for the review and commits. I will address those points in the forthcoming PR!
